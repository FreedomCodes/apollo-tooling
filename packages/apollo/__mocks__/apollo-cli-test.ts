import * as fs from "fs";
import Nock from "@fancy-test/nock";
import * as Test from "@oclif/test";
export { expect } from "@oclif/test";
import { mockConsole } from "heroku-cli-util";
// import { userInfo } from "os";
// import { vol } from "memfs";
// import { ufs } from "unionfs";

const time = label => {
  let start = +new Date();

  return {
    async run() {},
    finally() {
      console.log(`${label || "TIME"}:`, +new Date() - start, "ms");
    }
  };
};

const debug = fn => {
  return {
    async run() {
      fn();
    },
    finally() {}
  };
};

const deleteFolderRecursive = path => {
  if (fs.existsSync(path)) {
    fs.readdirSync(path).forEach(function(file, index) {
      var curPath = path + "/" + file;
      if (fs.lstatSync(curPath).isDirectory()) {
        // recurse
        deleteFolderRecursive(curPath);
      } else {
        // delete file
        fs.unlinkSync(curPath);
      }
    });
    fs.rmdirSync(path);
  }
};

const setupFS = (files: Record<string, string>) => {
  let dir;
  return {
    async run(ctx: any, ...rest) {
      // make a random remp dir & chdir into it
      dir = fs.mkdtempSync("__tmp__");
      process.chdir(dir);
      // fill the dir with `files`
      Object.keys(files).forEach(key => {
        fs.writeFileSync(key, files[key]);
      });
    },
    finally(ctx: any) {
      // const postFiles = fs.readdirSync("./");
      // delele all files
      // postFiles.forEach(key => {
      // fs.unlinkSync(key);
      // });
      // go up one level & delete the temp dir
      process.chdir("../");
      deleteFolderRecursive(dir);
      // fs.rmdirSync(dir);
    }
  };
};

export const test = Test.test
  .register("nock", Nock)
  .register("fs", setupFS)
  .register("timing", time)
  .register("debug", debug);

// it would be great to have an .expectStdout(out => expection)

mockConsole();
