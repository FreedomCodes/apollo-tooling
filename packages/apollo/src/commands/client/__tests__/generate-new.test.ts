// jest.mock("apollo-codegen-core/lib/localfs", () => {
//   return require("../../../../__mocks__/mockFs");
// });

// this is because of herkou-cli-utils hacky mocking system on their console logger
import { stdout, mockConsole } from "heroku-cli-util";
import * as path from "path";
import * as fs from "fs";
import { test as setup } from "apollo-cli-test";
import {
  introspectionQuery,
  print,
  execute,
  buildSchema,
  graphql
} from "graphql";
import gql from "graphql-tag";

const test = setup.do(() => mockConsole());

// helper function to resolve files from the actual filesystem
const resolveFiles = opts => {
  let files = {};
  Object.keys(opts).map(key => {
    files[key] = fs.readFileSync(path.resolve(__dirname, opts[key]), {
      encoding: "utf-8"
    });
  });

  return files;
};

const {
  graphQLSchema,
  simpleQuery,
  otherQuery,
  clientSideSchema,
  clientSideSchemaQuery,
  clientSideOnlySchema,
  clientSideOnlyQuery
} = resolveFiles({
  graphQLSchema: "../../service/__tests__/fixtures/schema.graphql",
  simpleQuery: "./fixtures/simpleQuery.graphql",
  otherQuery: "./fixtures/otherQuery.graphql",
  clientSideSchema: "./fixtures/clientSideSchema.graphql",
  clientSideSchemaQuery: "./fixtures/clientSideSchemaQuery.graphql",
  clientSideOnlySchema: "./fixtures/clientSideOnlySchema.graphql",
  clientSideOnlyQuery: "./fixtures/clientSideOnlyQuery.graphql"
});

const fullSchema = execute(buildSchema(graphQLSchema), gql(introspectionQuery))
  .data;

const clientSideSchemaTag = `
  gql\`
  ${clientSideSchema}
  \`
`;

const serverSideSchemaTag = `
  gql\`
  type Query {
    hello: String!
    serverSideField: ServerField!
  }

  type ServerField {
    serverData: String!
  }

  type RemovedField {
    id: ID!
    name: RemovedType
  }

  type RemovedType {
    fieldName: String
  }
  \`
  `;

describe("client:codegen", () => {
  test
    .fs({
      "./schema.json": JSON.stringify(fullSchema.__schema),
      "./queryOne.graphql": simpleQuery.toString(),
      "./apollo.config.js": `
        module.exports = {
          client: {
            includes: ["./**.graphql"],
            service: {
              name: "my-service-name",
              localSchemaFile: "./schema.json"
            }
          }
        }
      `
    })
    .command([
      "client:codegen",
      "API.swift",
      "--config=apollo.config.js",
      "--target=swift"
    ])
    .it("writes swift types from local schema", () => {
      expect(fs.readFileSync("API.swift").toString()).toMatchSnapshot();
    });

  test
    .fs({
      "schema.graphql": graphQLSchema,
      "queryOne.graphql": simpleQuery.toString(),
      "apollo.config.js": `
        module.exports = {
          client: {
            includes: ["./queryOne.graphql"],
            service: {
              name: "my-service-name",
              localSchemaFile: "./schema.graphql"
            }
          }
        }
      `
    })
    .command([
      "client:codegen",
      "API.swift",
      "--config=apollo.config.js",
      "--target=swift"
    ])
    .it("writes swift types from local schema in a graphql file", () => {
      expect(fs.readFileSync("API.swift").toString()).toMatchSnapshot();
    });

  test
    .fs({
      "schema.json": JSON.stringify(fullSchema.__schema),
      "queryOne.graphql": simpleQuery.toString(),
      "apollo.config.js": `
        module.exports = {
          client: {
            includes: ["./queryOne.graphql"],
            service: {
              name: "my-service-name",
              localSchemaFile: "./schema.json"
            }
          }
        }
      `
    })
    .command([
      "client:codegen",
      "API.swift",
      "--target=swift",
      "--operationIdsPath=myOperationIDs.json"
    ])
    .it("generates operation IDs for swift files when flag is set", () => {
      expect(
        fs.readFileSync("myOperationIDs.json").toString()
      ).toMatchSnapshot();
    });

  // FIXME: doesn't write anything significant to snapshot
  test
    .skip()
    .fs({
      "schema.json": JSON.stringify(fullSchema.__schema),
      "queryOne.graphql": simpleQuery.toString(),
      "queryTwo.graphql": otherQuery.toString(),
      "apollo.config.js": `
        module.exports = {
          client: {
            includes: ["./**.graphql"],
            service: {
              name: "my-service-name",
              localSchemaFile: "./schema.json"
            }
          }
        }
      `
    })
    .command([
      "client:codegen",
      "--only=queryTwo.graphql",
      "--target=swift",
      "outDirectory"
    ])
    .it("handles only flag for Swift target", () => {
      const [filePath] = fs.readdirSync("./outDirectory");
      const file = fs.readFileSync(`./outDirectory/${filePath}`).toString();

      // expect(file).toMatchSnapshot();
    });

  test
    .fs({
      "schema.json": JSON.stringify(fullSchema.__schema),
      "queryOne.graphql": simpleQuery.toString(),
      "apollo.config.js": `
        module.exports = {
          client: {
            includes: ["./**.graphql"],
            service: {
              name: "my-service-name",
              localSchemaFile: "./schema.json"
            }
          }
        }
      `
    })
    .command(["client:codegen", "--target=scala", "API.scala"])
    .it("writes types for scala", () => {
      expect(fs.readFileSync("API.scala").toString()).toMatchSnapshot();
    });

  test
    .fs({
      "schema.json": JSON.stringify(fullSchema.__schema),
      "queryOne.graphql": simpleQuery.toString(),
      "apollo.config.js": `
        module.exports = {
          client: {
            includes: ["./**.graphql"],
            service: {
              name: "my-service-name",
              localSchemaFile: "./schema.json"
            }
          }
        }
      `
    })
    .command([
      "client:codegen",
      "--target=typescript",
      "--outputFlat",
      "API.ts"
    ])
    .it("writes types for typescript", () => {
      expect(fs.readFileSync("API.ts").toString()).toMatchSnapshot();
    });

  test
    .fs({
      "schema.json": JSON.stringify(fullSchema.__schema),
      "clientSideSchema.graphql": clientSideSchema.toString(),
      "clientSideSchemaQuery.graphql": clientSideSchemaQuery.toString(),
      "apollo.config.js": `
        module.exports = {
          client: {
            includes: ["./**.graphql"],
            service: {
              name: "my-service-name",
              localSchemaFile: "./schema.json"
            }
          }
        }
      `
    })
    .command([
      "client:codegen",
      "--target=typescript",
      "--outputFlat",
      "API.ts"
    ])
    .it("writes typescript types for query with client-side data", () => {
      expect(fs.readFileSync("API.ts").toString()).toMatchSnapshot();
    });
});
