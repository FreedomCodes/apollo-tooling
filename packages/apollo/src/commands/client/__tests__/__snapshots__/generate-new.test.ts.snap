// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client:codegen generates operation IDs for swift files when flag is set 1`] = `
"{
  \\"4de1e386589b0853a102a87a3f21ca25266116517e59c1e8be9442e2571f00bc\\": {
    \\"name\\": \\"SimpleQuery\\",
    \\"source\\": \\"query SimpleQuery {\\\\n  hello\\\\n}\\"
  }
}"
`;

exports[`client:codegen writes swift types from local schema 1`] = `
"//  This file was automatically generated and should not be edited.

import Apollo

public final class SimpleQueryQuery: GraphQLQuery {
  public let operationDefinition =
    \\"query SimpleQuery {\\\\n  hello\\\\n}\\"

  public init() {
  }

  public struct Data: GraphQLSelectionSet {
    public static let possibleTypes = [\\"Query\\"]

    public static let selections: [GraphQLSelection] = [
      GraphQLField(\\"hello\\", type: .nonNull(.scalar(String.self))),
    ]

    public private(set) var resultMap: ResultMap

    public init(unsafeResultMap: ResultMap) {
      self.resultMap = unsafeResultMap
    }

    public init(hello: String) {
      self.init(unsafeResultMap: [\\"__typename\\": \\"Query\\", \\"hello\\": hello])
    }

    public var hello: String {
      get {
        return resultMap[\\"hello\\"]! as! String
      }
      set {
        resultMap.updateValue(newValue, forKey: \\"hello\\")
      }
    }
  }
}"
`;

exports[`client:codegen writes swift types from local schema in a graphql file 1`] = `
"//  This file was automatically generated and should not be edited.

import Apollo

public final class SimpleQueryQuery: GraphQLQuery {
  public let operationDefinition =
    \\"query SimpleQuery {\\\\n  hello\\\\n}\\"

  public init() {
  }

  public struct Data: GraphQLSelectionSet {
    public static let possibleTypes = [\\"Query\\"]

    public static let selections: [GraphQLSelection] = [
      GraphQLField(\\"hello\\", type: .nonNull(.scalar(String.self))),
    ]

    public private(set) var resultMap: ResultMap

    public init(unsafeResultMap: ResultMap) {
      self.resultMap = unsafeResultMap
    }

    public init(hello: String) {
      self.init(unsafeResultMap: [\\"__typename\\": \\"Query\\", \\"hello\\": hello])
    }

    public var hello: String {
      get {
        return resultMap[\\"hello\\"]! as! String
      }
      set {
        resultMap.updateValue(newValue, forKey: \\"hello\\")
      }
    }
  }
}"
`;

exports[`client:codegen writes types for scala 1`] = `
"//  This file was automatically generated and should not be edited.


object SimpleQueryQuery extends com.apollographql.scalajs.GraphQLQuery {
  val operationString =
    \\"query SimpleQuery {\\" +
    \\"  hello\\" +
    \\"}\\"
  val operation = com.apollographql.scalajs.gql(operationString)

  type Variables = Unit

  case class Data(hello: String) {
  }

  object Data {
    val possibleTypes = scala.collection.Set(\\"Query\\")
  }
}"
`;

exports[`client:codegen writes types for typescript 1`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export interface SimpleQuery {
  hello: string;
}

/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//==============================================================
"
`;

exports[`client:codegen writes typescript types for query with client-side data 1`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export interface SimpleQuery_complexLocalState {
  __typename: \\"LocalType\\";
  someData: string;
}

export interface SimpleQuery_serverSideField {
  __typename: \\"ServerField\\";
  serverData: string;
  addedLocalData: string;
}

export interface SimpleQuery_localServerModel {
  __typename: \\"ServerField\\";
  serverData: string;
}

export interface SimpleQuery {
  hello: string;
  localState: string;
  complexLocalState: SimpleQuery_complexLocalState;
  serverSideField: SimpleQuery_serverSideField;
  localServerModel: SimpleQuery_localServerModel;
}

/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//==============================================================
"
`;
